Adding GSOC report - kirtanchandak (#2)
openfoodfacts/gsoc-reports: 2025/kirtanchandak/kirtanchandak-gsoc-2025-report.md

Reviewer: remember to add a reference to this file in the global README.md

Author: resolved

- **Property Management**: Create tools for managing properties and their values
- **User Role Management**: Develop systems for managing user roles within the folksonomy engine
- **Exploration Tools**: Build tools for exploring folksonomy properties and product properties
- **Content Moderation**: Implement bulk operations for fighting vandalism and managing data quality

## Pre GSOC -

#### **Allow Searching Keys**

Implemented the ability to search for keys within the folksonomy API, making it easier for users to find relevant properties quickly.

- **Issue:** https://github.com/openfoodfacts/folksonomy_api/issues/19
- **PR:** https://github.com/openfoodfacts/folksonomy_api/pull/263

#### **Add Data to Local Database**

Added support for inserting data directly into the local folksonomy database. This makes it easier for contributors to test changes and add sample data without affecting the production environment.

- **Slack Discussion:** https://openfoodfacts.slack.com/archives/C01M1T2E8L8/p1741101888486549
- **PR:** https://github.com/openfoodfacts/folksonomy_api/pull/252

#### **Enable Getting Multiple Properties at Once**

> Enhanced the folksonomy API to allow fetching more than one property in a single request.

Reviewer: Explain the purpose of doing that (optimize performance to encourage for third party tools to use folksonomy engine as a database)

Author: resolved

- **Issue:** https://github.com/openfoodfacts/folksonomy_api/issues/236
- **PR:** https://github.com/openfoodfacts/folksonomy_api/pull/244

#### **Add Typing for OpenAPI Response for `/keys`**

Improved the folksonomy API's OpenAPI documentation by adding proper typing for the `/keys` endpoint response.

- **Issue:** https://github.com/openfoodfacts/folksonomy_api/issues/149
- **PR:** https://github.com/openfoodfacts/folksonomy_api/pull/230

## Feature Work Delivered For Google Summer of Code 2025

### **Folksonomy Editor Integration**

> Integrated existing `folksonomy-editor` webcomponent from the openfoodfacts-webcomponents into OpenFoodFacts. This work involved ensuring compatibility, fixing integration bugs, and making sure the components work seamlessly within the folksonomy interface. The integration improves code reusability and maintains consistency across different Open Food Facts applications.

Reviewer: You can add that contrary to previous implementations it uses the authenticatino token, making requests cheaper. It was not trivial to implement due to some CORS issues previously unseen in the API.

<p align="center">
  <img src="./assets/folksonomy-editor.png" width="700" />
</p>

**Relevant PRs:**

- https://github.com/openfoodfacts/openfoodfacts-server/pull/12136

### **Mobile Suggestion System**

Implemented a comprehensive suggestion system that works effectively on mobile devices. The system uses the folksonomy API endpoints, particularly the `/values/{k}` endpoint.

Documentation - https://api.folksonomy.openfoodfacts.org/docs#/Keys%20%26%20Values/get_unique_values_values__k__get

**Relevant PRs:**

- https://github.com/openfoodfacts/folksonomy_api/pull/263

### **Property Generation Wizard**

Created a simple yet powerful property generation wizard that automatically creates corresponding wiki pages for new properties. This tool streamlines the process of adding new property types to the system and ensures proper documentation is maintained.

The wizard guides users through the process of defining new properties, validates the input, and automatically generates the necessary wiki documentation, making it easier for the community to understand and use new properties.

> <p align="center">
>   <video width="700" controls>
>     <source src="./assets/folksonomy-wizard.mp4" type="video/mp4">
>     Your browser does not support the video tag.
>   </video>
> </p>

Reviewer: Beware that it does not display at all on Github, so I would add also add a link to the video.
```suggestion
<p align="center">
  <video width="700" controls>
    <source src="./assets/folksonomy-wizard.mp4" type="video/mp4">
    Your browser does not support the video tag.
  </video>
  [video showing the wizard in action](./assets/folksonomy-wizard.mp4)
</p>
```

**Relevant PRs:**

- https://github.com/openfoodfacts/openfoodfacts-webcomponents/pull/200

### **User Role Management System**

> Developed a comprehensive user role management system for the folksonomy engine. This allows us to perform operations like bulk deletions or property modifications are restricted to trusted contributors.

Reviewer: ```suggestion
Developed a comprehensive user role management system for the folksonomy engine. This allows us to restrict operations like bulk deletions or property modifications to trusted contributors.
```

Author: resolved

The role management system includes different user levels with appropriate permissions - `user`, `moderator` and `admin` helping maintain data quality while allowing community participation.

Implemented API for checking user role `/user/me` on OpenFoodFacts -

Documentation:

- https://api.folksonomy.openfoodfacts.org/docs#/default/get_user_info_user_me_get

**Relevant PRs:**

- https://github.com/openfoodfacts/folksonomy_api/pull/295

### **Property Exploration Tools**

### WebComponents:

Built comprehensive exploration tools for both folksonomy properties and folksonomy product properties. These tools allow users to browse, search, and analyze the existing properties in the system, making it easier to understand what data is available and how it's being used.

The exploration tools include filtering capabilities, download CSV, and action like delete, rename for moderators and admins.

**Relevant PRs:**

- https://github.com/openfoodfacts/openfoodfacts-webcomponents/pull/181
- https://github.com/openfoodfacts/openfoodfacts-webcomponents/pull/182

### Integration

Integrated these WebComponents into the Open Food Facts server, making the new tools available to all users.

> #### Folksonomy Product Properties -

Reviewer: ```suggestion
#### Folksonomy Product Properties
```

<p align="center">
  <img src="./assets/folksonomy-product-properties.png" width="700" />
</p>

> ### Folksonomy Properties -

Reviewer: The title level is important, because I was a bit lost otherwise !
```suggestion
#### Folksonomy Properties
```

<p align="center">
  <img src="./assets/folksonomy-properties.png" width="700" />
</p>

**Relevant PRs:**

- https://github.com/openfoodfacts/openfoodfacts-server/pull/12148
- https://github.com/openfoodfacts/openfoodfacts-server/pull/12277

### **Bulk Property/Value Management APIs**

Implemented a robust bulk value management system designed to fight vandalism and maintain data quality. The system includes features for bulk removal, renaming, deletion, replacement, and clash detection of property values.

This tool is essential for maintaining the integrity of the folksonomy-engine, allowing administrators to quickly address data quality issues, remove spam or incorrect entries. The clash detection feature helps prevent conflicts when merging or modifying values.

> Documantation:

Reviewer: ```suggestion
Documentation:
```

- https://api.folksonomy.openfoodfacts.org/docs#/Admin%20-%20Property%20Management
- https://api.folksonomy.openfoodfacts.org/docs#/Admin%20-%20Value%20Management

**Relevant PRs:**

- https://github.com/openfoodfacts/folksonomy_api/pull/302

### **Bulk Property/Value Management UI**

> Implemented UI for renaming, deleting and checking conflicts pefore deleting for values and properties.

Reviewer: I'm reading quickly and find some typo, so I suggest doing a thorough review on your side to spot remaining typos.
```suggestion
Implemented UI for renaming, deleting and checking conflicts before deleting for values and properties.
```
